### PR TITLE
feat(core): generate sign in & sign out uri

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

feat(core): generate sign in & sign out URI

## Testing

UT Passed.

### CODE REVIEW

File: core/sign_in.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	unescapedQueries, unescapeQueryErr := url.QueryUnescape(queries.Encode())
+
+	if unescapeQueryErr != nil {
+		return "", unescapeQueryErr
+	}
+
+	return uri.String() + "?" + unescapedQueries, nil
+}
+
+func appendScopeIfNotExist(scopes []string, scope string) []string {
```
Comment (charIeszhao) [inline core/sign_in.go:68]: nit

```suggestion
func appendScopeIfNotExisted(scopes []string, scope string) []string {
```

File: core/sign_in_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		CodeChallenge:         "codeChallenge",
+		State:                 "state",
+		Scopes:                []string{"openid", "offline_access", "profile"},
+		Resources:             []string{"resourceA", "resourceB"},
+		Prompt:                "consent",
+	})
+
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (charIeszhao) [inline core/sign_in_test.go:23]: Noob question: why is this needed?

File: core/sign_in_test.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	if signInUri != expectedSignInUri {
+		t.Errorf("Expected %s, got %s", expectedSignInUri, signInUri)
+	}
+}
+
```
Comment (charIeszhao) [inline core/sign_in_test.go:96]: I think we should add a test case to verify when `prompt` is specified as `Login`

File: core/sign_in_test.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+func TestGenerateSignInUriShouldNotContainsResourcesIfNoResourcesAreProvided(t *testing.T) {
```
Comment (charIeszhao) [inline core/sign_in_test.go:97]:
```suggestion
func TestGenerateSignInUriShouldNotContainResourcesIfNoResourcesAreProvided(t *testing.T) {
```

**Review by charIeszhao — APPROVED** [review]
Looks good

File: core/sign_in_test.go (hunk 2 above)
Comment (xiaoyijun) [reply core/sign_in_test.go:23]: We need the test to fail right now when `GenerateSignInUri` returns an error.

File: core/sign_in_test.go (hunk 3 above)
Comment (xiaoyijun) [reply core/sign_in_test.go:96]: updated.
